### ISSUE TITLE
Many ReactorBuilder-style tests in twisted.internet.test do not declare their reactor requirements via requiredInterfaces

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5521]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#5521|
|Type|defect|
|Created|2012-03-11 16:48:20Z|
|Branch|https://github.com/twisted/twisted/tree/better-test-declarations-5521|

`ReactorBuilder.requiredInterfaces` exists to allow tests which require a particular reactor interface to be skipped when an attempt is made to run those tests against a reactor which does not support that interface.

However, many of our tests (for TCP, for UNIX, for UDP, for time) do not use this feature and will still try to run against a reactor that does not provided the necessary interface.

Attachments:

* [required-interfaces.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b11/b1157ca063ecdb8429831f299edceece940de6f1/216a113549103d442961fa6e64b3c007acb01b00.patch) (5687 bytes) - added by exarkun on 2012-03-11 16:55:23Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5521 5521
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_better_test_declarations_5521 branches/better-test-declarations-5521
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__tests tests
time__1331484500000000 1331484500000000
changetime__1342274697000000 1342274697000000
version__None None
owner__exarkun exarkun
cc__mwh
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

Attached is a patch that fixes the instances of this that I found.  I didn't pay particularly close attention when constructing it, I just added requirements as quick as I could to get the tests to skip against my reactor that is missing most interfaces.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [34734]) Branching to 'better-test-declarations-5521'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [34735]) Declare that IReactorTCP is required by these test cases.  Also, re-wrap docstrings to 80 columns and unify the IPv6 skip definitions.

refs [#5521](https://github.com/twisted/twisted/issues/5521)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [34736]) Make all the necessary declarations for these

refs [#5521](https://github.com/twisted/twisted/issues/5521)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [34737]) The same thing for IReactorTime

refs [#5521](https://github.com/twisted/twisted/issues/5521)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [34739]) I feel as though I've been here before.

refs [#5521](https://github.com/twisted/twisted/issues/5521)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [34742]) This is redundant now, I think

refs [#5521](https://github.com/twisted/twisted/issues/5521)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

Just what you'd expect, but overlooked before because every existing reactor already implemented all of these interfaces and no one developed new reactors very TDDly.  [Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/better-test-declarations-5521).

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_10"></a>|@mwhudson set owner to @exarkun|
|-|-|

This seems fine to me, although I have no way of verifying that it works.  It certainly looks like it should :-)  I don't understand the 

```
if ipv6Skip:
    skip = ipv6Skip
```

construction, isn't that equivalent to plain ```skip = ipv6Skip```.  If you have a good reason for that, and you're sure the few failures in buildbot are unrelated (they certainly look unrelated to me), please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

> isn't that equivalent to plain `skip = ipv6Skip`

Sadly it is not.  A class-level `skip = None` is different than not having a class-level `skip` attribute at all.  It's terrible, and probably represents a bug or misfeature in trial to be fixed.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set status to `closed`|
|-|-|

(In [34751]) Merge better-test-declarations-5521

Author: exarkun
Reviewer: mwhudson
Fixes: [#5521](https://github.com/twisted/twisted/issues/5521)

Make sure to declare which reactor interfaces are required by various `ReactorBuilder`-style
tests.
